parser,backupccl: added ALTER BACKUP SQL statement

DarrylWong opened the pull request:
This change adds the ALTER BACKUP statement. ALTER BACKUP allows for users to re-encrypt an already KMS encrypted backup with new KMS URIs by decrypting the ENCRYPTION-INFO file with at least one of the old KMS keys and re-encrypting it with the new KMS URIs. This new ENCRYPTION-INFO file is then added in addition to the old one as we never want to overwrite/delete backups.

Teaches other statements, RESTORE and SHOW BACKUP to look for these new KMS encryption files by looking for the passed in KMS URI in all ENCRYPTION-INFO files.

Adds tests for ALTER BACKUP for ALTER BACKUP itself and to see that RESTORE AND SHOW BACKUP properly read in these new encryption files.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75900)
<!-- Reviewable:end -->

Commit:
backupccl, parser: added ALTER BACKUP SQL statement grammar

This change adds the ALTER BACKUP statement. ALTER BACKUP allows for users to re-encrypt an already KMS encrypted backup with new KMS URIs by decrypting the ENCRYPTION-INFO file with at least one of the old KMS keys and re-encrypting it with the new KMS URIs. This new ENCRYPTION-INFO file is then added in addition to the old one as we never want to overwrite/delete backups.

Teaches other statements, RESTORE and SHOW BACKUP to look for these new KMS encryption files by looking for the passed in KMS URI in all ENCRYPTION-INFO files.

Adds tests for ALTER BACKUP for ALTER BACKUP itself and to see that RESTORE AND SHOW BACKUP properly read in these new encryption files.

Release note (sql change): Allow users to add NEW_KMS encryption keys to existing backups using:
ALTER BACKUP <backup_location> ADD NEW_KMS = <kms_uris> WITH OLD_KMS = <kms_uris>
ALTER BACKUP <subdir> IN <backup_collection> ADD NEW_KMS = <kms_uris> WITH OLD_KMS = <kms_uris>

The OLD_KMS must refer to at least one KMS URI that was previously used to encrypt the backup. Following successful completion of the ALTER BACKUP, subsequent backups, restore and show commands can use any of old or new KMS URIs to decrypt the backup.

adityamaru commented on pkg/sql/parser/sql.y:
```diff
@@ -4295,6 +4297,40 @@ explain_option_list:
+// KMS:
+//    "[kms_provider]://[kms_host]/[master_key_identifier]?[parameters]" : add new kms keys to backup
+alter_backup_stmt:
+  ALTER BACKUP list_of_string_or_placeholder_opt_list
```
Spoke to @DarrylWong offline, I think we should be doing something similar to what `SHOW BACKUP` does and only accept a single backup directory that points all the way to the `ENCRYPTION-INFO` file. In that case this should change to `string_or_placeholder`.

adityamaru reviewed (commented):
This is looking good! Did a first pass, I actually haven't looked at the `doAlterBackup` logic and tests since there are some structural changes we discussed offline wrt what path an `ALTER BACKUP` should accept, and the fact that we should be able to restore/show backup using any of the old and new keys that backup was encrypted with. I'll take a deeper look once you've made those changes!

rhu713 reviewed (commented):
The code overall looks good! I just have one comment + one nit.

DarrylWong commented on pkg/ccl/backupccl/alter_backup_planning.go:
```diff
@@ -0,0 +1,280 @@
+	// Look for all files in dest that start with "/ENCRYPTION-INFO" and return them.
+	err := dest.List(ctx, "", "", func(p string) error {
+		p = strings.TrimPrefix(p, "/")
+		if match, err := regexp.MatchString(backupEncryptionInfoFile+".*", p); err != nil {
```
It reads in files as /<year>/<timestamp>/ENCRYPTION-INFO (not exactly that but you get the point) so I had to split it by `\` and take the last slice. Hopefully that's still better than the regex solution.

DarrylWong commented on pkg/ccl/backupccl/alter_backup_planning.go:
```diff
@@ -0,0 +1,280 @@
+	return nil
+}
+
+func getNumEncryptionInfoFiles(ctx context.Context, dest cloud.ExternalStorage) (int, error) {
```
I changed the function but pretty sure your concern is still valid. Not sure the answer though so I'll let aditya chime in

adityamaru commented on pkg/ccl/backupccl/alter_backup_planning.go:
```diff
@@ -0,0 +1,265 @@
+	// Use the first one that works to decrypt the ENCRYPTION-INFO file(s).
+	var defaultKMSInfo *jobspb.BackupEncryptionOptions_KMSInfo
+	oldKMSFound := false
+	for i := len(oldKms) - 1; i >= 0; i-- {
```
we can probably loop the oldKMs slice in increasing order, and change this to range based for loop?

adityamaru commented on pkg/ccl/backupccl/alter_backup_planning.go:
```diff
@@ -0,0 +1,265 @@
+	oldKMSFound := false
+	for i := len(oldKms) - 1; i >= 0; i-- {
+		for _, encFile := range opts {
+			defaultKMSInfo, err = validateKMSURIsAgainstFullBackup([]string{oldKms[i]},
```
nifty, I like it. Just for my edification, we can't pass the entire oldKMS slice as the first param here cause it's possible it has URIs that weren't used to originally encrypt this backup, and we don't want to fail in that case? In other words oldKMS does not need to be a strict subset of the keys used to encrypt this backup?

DarrylWong replied:
Yes and I believe that is what the CMEK team wanted, to be able to just give many keys at once in case one was down or the wrong one.

adityamaru commented on pkg/ccl/backupccl/alter_backup_planning.go:
```diff
@@ -0,0 +1,265 @@
+
+	ioConf := baseStores.ExternalIOConf()
+
+	// Check that at least one of the old keys works on the backup.
```
nit: Check that at least one of the old keys has been used to encrypt the backup in the past.

adityamaru commented on pkg/ccl/backupccl/alter_backup_planning.go:
```diff
@@ -0,0 +1,265 @@
+				break
+			}
+		}
+		if oldKMSFound {
```
can we use `defaultKMSInfo != nil` here and below instead of the bool?

adityamaru commented on pkg/ccl/backupccl/alter_backup_planning.go:
```diff
@@ -0,0 +1,265 @@
+		}
+	}
+	if !oldKMSFound {
+		return fmt.Errorf("no key in old_kms matches the provided backup")
```
nit: return errors.New(...)

adityamaru commented on pkg/ccl/backupccl/alter_backup_planning.go:
```diff
@@ -0,0 +1,265 @@
+	var plaintextDataKey []byte
+	plaintextDataKey, err = getEncryptionKey(ctx, encryption, baseStores.Settings(),
+		baseStores.ExternalIOConf())
+
```
nit: delete newline between error and error checks

adityamaru commented on pkg/ccl/backupccl/alter_backup_planning.go:
```diff
@@ -0,0 +1,265 @@
+		masterKeyID, encryptedDataKey, err := getEncryptedDataKeyFromURI(ctx,
+			plaintextDataKey, kmsURI, kmsEnv)
+		if err != nil {
+			return err
```
maybe wrap this error so that its more informative when it bubble up. Something like:

return errors.Wrap(err, "failed to encrypt data key when adding new KMS")

adityamaru commented on pkg/ccl/backupccl/alter_backup_planning.go:
```diff
@@ -0,0 +1,265 @@
+	return files, err
+}
+
+func writeAlteredEncryptionInfo(
```
nit: rename to `writeNewEncryptionInfoToBackup`

can we just pass in the new file name from the doPlanHook method? That way we don't have to call `getEncryptionInfoFiles` again, cause we already do that when checking the oldKMSURIs.

adityamaru reviewed (commented):
It's getting there! did a round of comments on `ALTER BACKUP` that I thought I'd send out. I'll look at the tests and restore/show once you address these, and then I think we should be good to go.

adityamaru commented on pkg/sql/parser/sql.y:
```diff
@@ -4295,6 +4297,36 @@ explain_option_list:
+			NewKMSURI:  $9.stringOrPlaceholderOptList(),
+			OldKMSURI:  $13.stringOrPlaceholderOptList(),
+		}
+	}
```
See the new `ALTER CHANGEFEED` syntax for some reference.

rafiss reviewed (changes requested):
drive-by comment from sql-experience: please add tests for the parsing in `pkg/sql/parser/testdata` (probably make a new file called `alter_backup`)

run with: `make test PKG=./pkg/sql/parser TESTS='TestParseDatadriven'`
update the expected output with `make test PKG=./pkg/sql/parser TESTS='TestParseDatadriven' TESTFLAGS='-rewrite'`

rafiss reviewed (approved):
um nvm, i see the test now. the github diff didn't load that before :facepalm:

i'll leave the rest of the review to bulk-io!

adityamaru commented on pkg/ccl/backupccl/alter_backup_test.go:
```diff
@@ -0,0 +1,167 @@
+
+// TestAlterBackupShowBackup tests to see that show backup correctly recognizes the new encryption-info file
+// when SHOW BACKUP is called on an encrypted backup that was altered.
+func TestAlterBackupShowBackup(t *testing.T) {
```
can I suggest a slight restructure:

- In `TestAlterBackupRestore` you can keep the create table, backup, alter backup statements
- Then create t.Run() subtests for RESTORE and SHOW that use the common backup. This way you only do the setup once and the test is more readable.

I'd also like to see tests for:
- More than one NEW_KMS
- More than one OLD_KMS (with maybe one invalid, one valid)
- And a new kms == old kms maybe? It should just work but better to test it IMO.

I'll leave it up to you to decide which of these are easy enough to add in!

DarrylWong replied:
Done. Although I couldn't get the incremental backup tests to work in there since the extra files it makes was messing up some other tests. Let me know if you think this is an issue though, I can probably figure it out just wasn't sure if it was worth the effort to add in just so it could be in the same test.

adityamaru reviewed (commented):
Mostly nits, everything basically LGTM. I'll do one more round before stamping it!

adityamaru commented:
Comment on the commit message:

I think we should add a little more color about the implementation changes in this PR. Specifically, calling out that an `ALTER BACKUP` to re-encrypt a backup with new KMS URIs will result in a new ENCRYPTION-INFO file being written in the backup bucket. 
We should then mention that RESTORE, and SHOW BACKUP will check all the encryption info files to find the passed in KMS URI, and decrypt the backup.

The release note should also be written for the end user, I'd do something like:

Release note (sql change): Allow users to add NEW_KMS encryption keys to existing backups using:
`ALTER BACKUP <backup_location> ADD NEW_KMS = <kms_uris> WITH OLD_KMS = <kms_uris>`
`ALTER BACKUP <subdir> IN <backup_collection> ADD NEW_KMS = <kms_uris> WITH OLD_KMS = <kms_uris>`

The OLD_KMS must refer to atleast one KMS URI that was previously used to encrypt the backup. Following successful completion of the `ALTER BACKUP`, subsequent backups, restore and show commands can use any of old or new KMS URIs to decrypt the backup.

adityamaru commented on pkg/ccl/backupccl/manifest_handling.go:
```diff
@@ -1037,23 +1037,57 @@ func sanitizeLocalityKV(kv string) string {
+// readEncryptionOptions takes in a backup location and tries to find and return all encryption option files in the
+// backup. A backup normally only creates one encryption option file, but if the user uses ALTER BACKUP to add new
+// keys, a new encryption option file will be placed side by side with the old one. Since the old file is still valid,
+// as we never want to modify or delete an existing backup, we return both new and old files.
```
nit: we usually wrap comments to 70 characters, i think you can configure goland to do this

adityamaru replied:
here and the few others places you have long comments in this PR

adityamaru commented on pkg/ccl/backupccl/alter_backup_test.go:
```diff
@@ -0,0 +1,211 @@
+func getAWSEncryptionOption(
+	t *testing.T, regionEnvVariable string, keyIDEnvVariable string,
+) string {
+	uri, _ := getAWSKMSURI(t, regionEnvVariable, keyIDEnvVariable)
```
hmm we're ignoring the error returned here, what we actually want to do is skip the test if we see an error right? Which is what CI is going to do since it doesn't have these env variables configured?

DarrylWong replied:
getAWSKMSURI returns both a correct and incorrect URI, no errors, but I only need the correctURI which is why I'm ignoring the incorrect one. It already skips the test for you if the env variable isn't set, which is why you have to pass in t. Currently already skips the test if not set and seems to be passing Cl.

Only thing getAWSEncryptionOption does is wrap the correctURI in single quotes so I didn't have to do Sprintf and do it every single time.

adityamaru replied:
my bad, we have two `getAWSKMSURI` in the codebase, I was looking at the wrong one. Thanks for the explanation, my only remaining comment is that you've run these tests once with actual KMS keys.

DarrylWong replied:
Yup I've run them with keys and all seems to be passing!

adityamaru replied:
Nope, 🚢 it!

adityamaru reviewed (approved):
Thanks for chipping away at this. The formatting in sql.y still looks a little wonky in github but we can clean it up in a follow up if there is anything to be done.

The one concern I have is that the tests are swallowing the error if the required env vars aren't set. We should be skipping the test instead like we do for `aws_kms_test.go`. I also assume you have run configured AWS / your local env to run these tests for reals one time.

A stability work item would be to teach our unit test CI to run these tests instead of skipping them. Finally, we should probably tweak our roachtest to also use this alter syntax once it is checked in. This can be done in a follow-up PR.

DarrylWong commented:
bors r+

shermanCRL commented:
cc @livlobo @benbardin
